File: src/main/java/org/veupathdb/lib/blast/field/OutFile.java
package org.veupathdb.lib.blast.field;

import com.fasterxml.jackson.databind.JsonNode;
import org.veupathdb.lib.blast.util.DefaultingJSONValue;
import org.veupathdb.lib.blast.util.JSONConstructor;

public record OutFile(String value) implements DefaultingJSONValue
{
  public static final String DefaultValue = "-";

  @Override
  public boolean isDefault() {
    return value == null || DefaultValue.equals(value);
  }

  @Override
  public JsonNode toJSON() {
    return JSONConstructor.newText(value);
  }
}

File: src/main/kotlin/org/veupathdb/lib/blast/field/out-file.kt
package org.veupathdb.lib.blast.field

import com.fasterxml.jackson.databind.node.ObjectNode
import org.veupathdb.lib.blast.serial.BlastField


private const val DefaultOutFile = "-"
private const val KeyOutFile     = "-out"


internal fun ParseOutFile(js: ObjectNode) =
  OutFile(js[KeyOutFile]?.textValue() ?: DefaultOutFile)

@JvmInline
value class OutFile(val value: String = DefaultOutFile): BlastField {
  override val isDefault get() = value == DefaultOutFile && value.isNotBlank()

  override fun appendJson(js: ObjectNode) {
    if (!isDefault)
      js.put(KeyOutFile, value)
  }

  override fun appendCliSegment(cli: StringBuilder) {
    if (!isDefault)
      cli.append(' ')
        .append(KeyOutFile)
        .append(" '")
        .append(value)
        .append('\'')
  }

  override fun appendCliParts(cli: MutableList<String>) {
    if (!isDefault) {
      cli.add(KeyOutFile)
      cli.add(value)
    }
  }
}